src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,321 @@
+lemma _root_.singleton_mul_subgroup {H : subgroup G} {h : G} (hh : h ∈ H) :
+  {h} * (H : set G) = H :=
+begin
+  refine le_antisymm _ (λ h' hh', ⟨h, h⁻¹ * h', rfl, H.mul_mem (H.inv_mem hh) hh',
+    mul_inv_cancel_left h h'⟩),
+  rintros _ ⟨h, h', rfl : _ = _, hh', rfl⟩,
+  exact H.mul_mem hh hh',
+end
+
+lemma sub_doset {H K : subgroup G} {a b : G} (hb : b ∈ doset a H K) :
+  doset b H K = doset a H K :=
+begin
+  obtain ⟨_, k, ⟨h, a, hh, (rfl : _ = _), rfl⟩, hk, rfl⟩ := hb,
+  rw [doset, doset, ←set.singleton_mul_singleton, ←set.singleton_mul_singleton, mul_assoc,
+      mul_assoc, singleton_mul_subgroup hk, ←mul_assoc, ←mul_assoc, subgroup_mul_singleton hh],
+end
+
+lemma rel_iff {H K : subgroup G} {x y : G} :
+  double_coset_rel ↑H ↑K x y ↔ ∃ (a ∈ H) (b ∈ K), y = a * x * b :=
+iff.trans ⟨λ hxy, (congr_arg _ hxy).mpr (doset_subgroups_mem_self H K y),
+  λ hxy, (sub_doset hxy).symm⟩ doset_mem
+
+lemma left_bot_eq_left_group_rel (H : subgroup G) :
+  (double_coset_rel ↑(⊥ : subgroup G) ↑H) = (quotient_group.left_rel H).rel :=
+begin
+  show (double_coset_rel ↑(⊥ : subgroup G) ↑H) = (quotient_group.left_rel H).r,
+  ext,
+  rw rel_iff,
+  simp only [exists_prop, one_mul, subgroup.mem_bot, exists_eq_left, exists_prop, one_mul,
+  subgroup.mem_bot, exists_eq_left],
+  split,
+  rintro ⟨a, ha⟩,
+  rw [quotient_group.left_rel, ha.2],
+  simp only [ha.left, inv_mul_cancel_left],
+  rw quotient_group.left_rel,
+  dsimp only,
+  intro h,
+  use (⟨x⁻¹ * x_1, h⟩ : H),
+  simp only [h, mul_inv_cancel_left, eq_self_iff_true, and_self, subgroup.coe_mk],
+end
+
+lemma right_bot_eq_right_group_rel (H : subgroup G) :
+  (double_coset_rel ↑H ↑(⊥ : subgroup G)) = (quotient_group.right_rel H).rel :=
+begin
+  show (double_coset_rel ↑H ↑(⊥ : subgroup G)) = (quotient_group.right_rel H).r,
+  ext,
+  rw rel_iff,
+  simp only [exists_prop, mul_one, subgroup.mem_bot, exists_eq_left, subgroup.coe_bot] at *,
+  split,
+  rintro ⟨a, ha⟩,
+  rw [quotient_group.right_rel, ha.2],
+  simp only [ha.left, mul_one, mul_inv_cancel_right],
+  rw quotient_group.right_rel,
+  dsimp only,
+  intro h,
+  use (⟨x_1 * x⁻¹, h⟩ : H),
+  simp only [h, mul_one, eq_self_iff_true, and_self, subgroup.coe_mk, inv_mul_cancel_right],
+end
+
+lemma disjoint_sub (H K : subgroup G) (a b : G) (h : ¬ disjoint (doset a H K ) (doset b H K )) :
+  b ∈ doset a H K :=
+begin
+  rw set.not_disjoint_iff at h,
+  simp only [exists_prop, set_coe.exists, doset_mem, subgroup.mem_carrier, set_like.mem_coe,
+    subgroup.coe_mk] at *,
+  obtain  ⟨x, ⟨l, hl, r, hr, hrx⟩, y, hy, ⟨r', hr', hly⟩⟩ := h,
+  use y⁻¹ * l,
+  have hyl :  y⁻¹*l ∈ H, by {apply subgroup.mul_mem H (subgroup.inv_mem H hy) (hl), },
+  simp only [hyl, true_and],
+  use r*r'⁻¹,
+  have hrr' : r*r'⁻¹ ∈ K, by {apply subgroup.mul_mem K hr (subgroup.inv_mem K hr') ,},
+  simp [hrr'],
+  rw ←mul_assoc,
+  rw hly at hrx,
+  rw ← mul_inv_eq_iff_eq_mul,
+  simp,
+  rw [mul_assoc, mul_assoc],
+  simp_rw eq_inv_mul_iff_mul_eq,
+  simp [← mul_assoc, hrx],
```

Comment:
Here are some golfing tips:
1) You can sometimes use "rfl" in `obtain` and rintros` to do substitutions automatically. In this case, you can replace `hly` with `rfl` and avoid having to do `rw hly at hrx,` later on.
2) After the `obtain`, you have a big goal involving a bunch of exists and ands. Rather than using `use` and `simp` to break up the goal, it's easier to use refine. As a first attempt, writing `refine ⟨_, _, _, _, _⟩,` after the `obtain` splits the goal into five subgoals. But now you can start filling in some of the goals ahead of time. As a second attempt, writing `refine ⟨y⁻¹ * l, _, r * r'⁻¹, _, _⟩,` after the `obtain` splits the goal into three subgoals. Even better is `refine ⟨y⁻¹ * l, H.mul_mem (H.inv_mem hy) (hl), r * r'⁻¹, K.mul_mem hr (K.inv_mem hr'), _⟩,` which uses golfed versions of your first two proofs, and just leaves one goal remaining.
```suggestion
  obtain  ⟨x, ⟨l, hl, r, hr, hrx⟩, y, hy, ⟨r', hr', rfl⟩⟩ := h,
  refine ⟨y⁻¹ * l, H.mul_mem (H.inv_mem hy) (hl), r * r'⁻¹, K.mul_mem hr (K.inv_mem hr'), _⟩,
  rwa [mul_assoc, mul_assoc, eq_inv_mul_iff_mul_eq, ←mul_assoc, ←mul_assoc, eq_mul_inv_iff_mul_eq],
```

Reply:
Oh that very useful! Thank you! I'll see if I can figure out how to use this to do some golfing. This is one thing I'd very much like to get better at, so tips like this are quite useful.

Reply:
So I tried some of this and it did help, so thank you! That said, I expect more golfing is possible.